src/main/java/com/snowflake/kafka/connector/internal/SnowflakeInternalStage.java

```diff
@@ -65,72 +106,139 @@ public SnowflakeInternalStage(
+        storageInfoCache.get(stageName).fileTransferMetadata;
+    // Set filename to be uploaded
+    // This set is not useful in GCS since there is a bug in JDBC which doesnt use destFileName.
+    fileTransferMetadata.setPresignedUrlFileName(fullFilePath);
 
+    byte[] dataBytes = data.getBytes(StandardCharsets.UTF_8);
+    InputStream inStream = new ByteArrayInputStream(dataBytes);
+
+    // This uploadWithoutConnection api cannot handle expired credentials very well.
+    // Need to prevent passing expired credential to it.
+    try {
+      SnowflakeFileTransferAgent.uploadWithoutConnection(
+          SnowflakeFileTransferConfig.Builder.newInstance()
+              .setSnowflakeFileTransferMetadata(fileTransferMetadata)
+              .setUploadStream(inStream)
+              .setRequireCompress(true)
+              .setOcspMode(OCSPMode.FAIL_OPEN)
+              .setProxyProperties(proxyProperties)
+              .build());
     } catch (Exception e) {
-      logWarn("Caught exception in putWithCache for fileName:{}", e.getMessage(), fullFilePath);
-      throw SnowflakeErrors.ERROR_5018.getException(e);
+      // If this api encounters error, invalidate the cached credentials
+      // Caller will retry this function
+      logWarn(
+          "uploadWithoutConnection encountered an exception:{} for filePath:{} in Storage:{}",
+          e.getMessage(),
+          fullFilePath,
+          stageType);
+      storageInfoCache.remove(stageName);
+      throw SnowflakeErrors.ERROR_5018.getException(e.getMessage());
+    }
+  }
+
+  /**
+   * Check if credentials are valid before calling uploadWithoutConnection API. Valid if they are
+   * not null and last fetch time was within {@link #expirationTimeMillis} For GCS, this will be
+   * false and we always refresh the credentials.
+   *
+   * @param credential to check the validity for
+   * @param stageType stageType of the stage fetched from JDBC and at least one GS call
+   * @return true if we can reuse the credentials, false in case of GCS
+   */
+  private boolean isCredentialValid(
+      SnowflakeMetadataWithExpiration credential, final StageInfo.StageType stageType) {
+
+    if (stageType == StageInfo.StageType.GCS) {
+      return false;
     }
+    // Key is cached and not expired
+    return credential != null
+        && System.currentTimeMillis() - credential.timestampLastCredentialFetchMillis
+            < expirationTimeMillis;
   }
 
-  private boolean isCredentialValid(SnowflakeMetadataWithExpiration credential) {
-    // Kay is cached and not expired
-    return credential != null && System.currentTimeMillis() - credential.timestamp < expirationTime;
+  @VisibleForTesting
+  protected void refreshCredentials(
+      final String stageName, final StageInfo.StageType stageType, final String fullFilePath)
+      throws SnowflakeSQLException {
+    String putCommandToFetchMetadata;
+    if (stageType == StageInfo.StageType.GCS) {
```

Comment:
Can this be a helper method?  We can make this code can be free of stage type/cloud details

Reply:
yes, makes sense. lemme decouple.